Mobile messaging

wzieba opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #34544.
Project thread (internal): pe5pgL-qo-p2

This PR adds marketing messages to "New Order" mails. The following algorithm selects the messages: 

<details>
<summary>Old copy selection logic (iteration 2)</summary>
<img src=https://user-images.githubusercontent.com/5845095/189615267-b6af7e77-5da6-4f07-87e3-36c1a429ed33.png>
</details>
Up-to-date copy selection logic (iteration 3)
<img width="1070" alt="image" src="https://user-images.githubusercontent.com/5845095/190224439-69ddc6a1-7a02-4a8b-9bc1-39e54bbdb7e1.png">


### How to test the changes in this Pull Request:

#### TC1: Store and order eligible for IPP
1. Assert that store is eligible for IPP (is based in the US with USD currency or in CA with CAD)
2. Create a new, IPP-eligible Order (empty order in Processing state is okay. All requirements: PdfdoF-V-p2)
3. Assert that received "New Order" mail has the "Collect payment easily (...)" message in the footer.

#### TC2: Order not eligible for IPP
1. With the same site as above, create a new non-eligible IPP Order (e.g. set payment method Paypal or add subscription products)
2. Assert that received "New Order" mail has the "Manage the order (...)" message in the footer.

#### TC3: Store not eligible for IPP
1. Change the store currency to other than USD (if the store was the US) or CAD (if the store was the CA).
2. Create a new, IPP-eligible Order
3. Assert that received "New Order" mail has the "Manage the order (...)" message in the footer.

#### TC4: No Jetpack
1. Revert the store's currency change from TC3 if needed.
2. Disable Jetpack
3. Create a new, IPP-eligible Order
4. Assert that received the "Process your orders on the go (...) message in the footer.


### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

wzieba committed:
feat: add mobile deep link footer

wzieba committed:
docs: add changelog entry

wzieba committed:
style: rename messaging handler

wzieba committed:
fix: do not crash if only ios/android available

wzieba committed:
fix: prepare translateable URL

wzieba committed:
fix: hook into email_footer action

wzieba committed:
fix: remove mobile apps usage condition

wzieba committed:
feat: update link to have 'order' segment

wzieba committed:
Revert "fix: remove mobile apps usage condition"

This reverts commit 1c3827f917fb516ba11c83ab059637fd2f77a39e.

wzieba committed:
feat: set last  mobile app usage interval to 30 days

wzieba committed:
feat: add new messages and conditions

wzieba commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+			$used_app_in_last_month = $last_mobile_used->diff( $now )->days <= self::OPEN_ORDER_INTERVAL_DAYS;
+			$has_jetpack            = $blog_id !== null;
+
+			if ( $used_app_in_last_month && $has_jetpack ) {
```
This logic block is WIP, see (internal):pe5pgL-EE-p2

wzieba commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+	 */
+	private static function is_order_in_person_payment_eligible( WC_Order $order ): bool {
+		return in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true ) &&
+			   in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true ) && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
+			   self::is_store_in_person_payment_eligible() && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
+			   ! $order->is_paid() && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
```
I have no idea how to fix this: neither `phpcbf` nor PHPStorm WordPress code style doesn't fix those issues. I also couldn't fix this manually (or maybe I don't understand the message).

coreymckrill replied:
I think it means that each condition is using individual space characters in order to line up with the condition on the line above it, but the code style rules say that all indenting should be done with tabs, not spaces.

Honestly, though, the logic is a little hard to follow with so many conditions. My suggestion would be to assign each of those conditions to a separate boolean variable and then just check that they're all true at the end. Something like this:

```php
$has_status = in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true );
$has_method = in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true );
$store_is_eligible = self::is_store_in_person_payment_eligible();
$order_is_not_paid = ! $order->is_paid();
$order_is_not_refunded = ! empty( $order->get_refunds() );

return $has_status && $has_method && $store_is_eligible && $order_is_not_paid && $order_is_not_refunded;
```

wzieba commented on plugins/woocommerce/templates/emails/email-mobile-messaging.php:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @see https://docs.woocommerce.com/document/template-structure/
+ * @package WooCommerce\Templates\Emails
+ * @version 6.9.0
```
I'll change it to 7.0 or 7.1.

coreymckrill replied:
Might be best to leave it as `x.x.x` or something for now, so we don't forget about it later.

wzieba commented on plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @package WooCommerce\Tests
+ */
+class WC_Mobile_Messaging_Handler_Test extends WC_Unit_Test_Case {
```
I'll add more unit tests later, ideally to fully cover `WC_Mobile_Messaging_Handler` - I didn't do it yet as the SUT class can heavily change at this point.

wzieba commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+	 *
+	 * @return string|null
+	 */
+	public static function prepare_mobile_message(
```
I have concerns about "staticness" of this class. Is it acceptable? Or should I (somehow) inject `WCTracker` and make it not static? Or it's not needed?

coreymckrill replied:
I think static methods are ok in this case. One thing that you're doing that isn't really used anywhere else in WooCommerce (as far as I know) is specifying the return type. While I think that's technically compatible with the minimum version of PHP required by WC, it's a lot stricter than usual. But we can check with someone on the core team about it.

coreymckrill replied:
It looks like there are a few other places in the codebase (all within the `src` directory) that use return types, so I guess this is fine.

wzieba commented on plugins/woocommerce/changelog/Mobile deep link email footer:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: enhancement
+
+Adds mobile deep link email footer
```
- [x] To change later.

wzieba committed:
Merge branch 'trunk' into feature/footer_with_mobile_deep_link

wzieba committed:
style: use yoda conditions

wzieba commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+	 *
+	 * @return true if order is eligible, false otherwise
+	 */
+	private static function is_order_in_person_payment_eligible( WC_Order $order ): bool {
```
Logic was taken from internal documentation: PdfdoF-V-p2 except for one condition

- [x] TODO: add "Order does NOT have any subscription products" condition

wzieba commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+	private static function is_order_in_person_payment_eligible( WC_Order $order ): bool {
+		return in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true ) &&
+			   in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true ) && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
+			   self::is_store_in_person_payment_eligible() && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
```
This condition is duplicated (at this point we already checked if store is IPP eligible) but I find this check needed to make sure that `is_order_in_person_payment_eligible` function does exactly what it says it does - checks all conditions.

I believe it's so minor that not even measurable in terms of performance so decided to leave it.

coreymckrill replied:
(See my other comment on the next line down)

wzieba reviewed (commented):
Hi @coreymckrill 👋 

I've moved to the original repository and added the improvements you mentioned in the first review - I hope I understood them correctly.

Also, I've added a few lines of new code as the requirements changed.

Please feel free to take a look at the code and share if things go in the right direction. Thanks!

Ps. I don't understand why tests are failing 🤔 they seem to run out of memory but I'm not sure if this is something I could affect with code changes in this PR.

wzieba commented on plugins/woocommerce/includes/emails/class-wc-email.php:
```diff
@@ -256,6 +256,25 @@ public function __construct() {
 
 		add_action( 'phpmailer_init', array( $this, 'handle_multipart' ) );
 		add_action( 'woocommerce_update_options_email_' . $this->id, array( $this, 'process_admin_options' ) );
+		add_action( 'woocommerce_email_footer', array( $this, 'mobile_messaging' ), 9 ); // Run before the default email footer.
```
I'm not sure why there's `priority 9`. Should `mobile_messaging` be executed before `woocommerce_email_footer` (as your comment says) or I didn't understand it?

<img width="333" alt="image" src="https://user-images.githubusercontent.com/5845095/187238499-c219e0ef-d2dc-4d9f-941c-6dd31cf4bcf6.png">

coreymckrill replied:
Yeah, lower priority number means it will happen sooner. The default priority is `10`. We want this message to be output before the footer.

coreymckrill commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%1$s">Accept payments</a> with a card reader in our mobile app\.</br><a href="%2$s">Learn more about In-Person Payments\.</a>',
+					'woocommerce'
+				)
```
1. I would probably move the `translators` comment down to a line above the beginning of the string, to make sure it gets captured by the translation string parser. (Same with the other two instances farther down)
    ```
	__(
		/* translators: %s: Email link */
		'<a href="%1$s">Accept payments</a>...'
    ```
2. I don't think the period characters need to be escaped in this string.
3. `</br>` should be `<br />` instead.

wzieba replied:
> I would probably move the translators comment down to a line above the beginning of the string, to make sure it gets captured by the translation string parser.

I've tried doing this:

<img width="1382" alt="image" src="https://user-images.githubusercontent.com/5845095/187388453-874996c4-85ca-4706-aeb8-74d856715825.png">

and phpcs returns:

```
| ERROR | A gettext call containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
```

so I'm not sure we can do this.

coreymckrill replied:
Oh interesting. I assumed the translators comment needed to be adjacent to the actual text string, but I guess it needs to be adjacent to the gettext function ( `__()` ) instead. So, you're already doing it right.

coreymckrill commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,207 @@
+				)
+			),
+			esc_url( $deep_link_url ),
+			esc_url( 'https://woocommerce.com/in-person-payments/' )
```
This doesn't need the `esc_url` wrapper since it's a static string instead of a variable.

wzieba committed:
style: extract conditions to variables

wzieba committed:
docs: update email template comment to not contain release version

wzieba committed:
fix: satisfy updated method signature with order instead id

wzieba committed:
fix: do not escape plain urls

wzieba committed:
fix: fix messages formatting

wzieba committed:
feat: check if order has no subscription products

wzieba committed:
fix: do not crash for WCS_Email_New_Renewal_Order

wzieba committed:
fix: conditions making order eligible for ipp

wzieba committed:
fix: do not crash when no mobile app usage reports

wzieba committed:
test: add unit tests for conditions to select the correct mobile message

wzieba committed:
test: add unit tests for status, payment methods and paid conditions

wzieba committed:
ci(temp): disable canceling unit tests jobs

wzieba committed:
revert: unwanted code style changes

wzieba committed:
Merge branch 'trunk' into feature/footer_with_mobile_deep_link

wzieba commented:
I've extracted some IPP-specific methods to a separate class to make it easier to unit-test and to reduce `WC_Mobile_Messaging_Handler` responsibilities.

But I see I'm making some mistakes with the unit tests. It seems that those unit tests are not independent - `WC_Ipp_Functions_Test` is passing locally, but I see it's failing on CI (when all tests are executed). Do you have an idea or suggestion on how I could make those unit tests independent?

wzieba committed:
fix: do not check for store ipp eligibility when checking order

wzieba committed:
fix: setting country in tests

wzieba committed:
feat: update copy

wzieba committed:
Merge branch 'trunk' into feature/footer_with_mobile_deep_link

wzieba committed:
test: improve test suite to revert initial country/currency state. Improve not ipp-eligible store test.

coreymckrill commented:
@wzieba 

> I think that unit test jobs are canceled because performance tests for placing an order fail.

I'm not totally understanding the issue. Can you point me to the tests that are failing?

> It's probably expected - we started to do more operations to prepare a new order email. Should we increase the expected baseline times? Or should we approach it differently, WDYT?

If the failures are actually directly related to these changes then I think we should increase the baselines, but I'm not convinced yet that that's what's going on...

coreymckrill commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,171 @@
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%1$s">Collect payments easily</a> from your customers anywhere with our mobile app.',
```
```suggestion
					'<a href="%s">Collect payments easily</a> from your customers anywhere with our mobile app.',
```

The enumerated token syntax isn't needed since there's only one token in the string.

wzieba commented:
> I'm not totally understanding the issue. Can you point me to the tests that are failing?

I think it happened e.g. [here](https://github.com/woocommerce/woocommerce/runs/8090449897?check_suite_focus=true#step:8:1192) but let me re-enable performance tests to see how things will work now.

wzieba committed:
Revert "ci(temp): disable canceling unit tests jobs"

This reverts commit ec55428ec09b1801d84f44c9c208e58a38ad3204.

wzieba committed:
fix: do not use enumerated syntax when there's only one token

Co-authored-by: Corey McKrill <916023+coreymckrill@users.noreply.github.com>

wzieba committed:
Merge branch 'trunk' into feature/footer_with_mobile_deep_link

wzieba committed:
docs: update changelog entry

wzieba committed:
docs: update phpdoc versions

coreymckrill commented on plugins/woocommerce/includes/wc-ipp-functions.php:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @param WC_Order $order order that the conditions are checked for.
+ *
+ * @return true if order is eligible, false otherwise
```
```suggestion
 * @return bool True if order is eligible, false otherwise.
```

The `@return` type should be shown as `bool`. Right now my IDE interprets this as the function always returning specifically `true`. Same goes for the other functions in this file.

coreymckrill commented on plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php:
```diff
@@ -0,0 +1,151 @@
+	/**
+	 * Saves values of initial country and currency before running test suite.
+	 */
+	public static function setUpBeforeClass(): void {
```
```suggestion
	public static function wpSetUpBeforeClass(): void {
```

Probably best to use the specific WP child method for this instead of the generic phpunit one.

coreymckrill commented on plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php:
```diff
@@ -0,0 +1,151 @@
+	/**
+	 * Restores initial values of country and currency after running test suite.
+	 */
+	public static function tearDownAfterClass(): void {
```
```suggestion
	public static function wpTearDownAfterClass(): void {
```

Probably best to use the specific WP child method for this instead of the generic phpunit one.

coreymckrill commented on plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php:
```diff
@@ -0,0 +1,151 @@
+		$this->assertEquals(
+			'<a href="https://woocommerce.com/mobile/order?blog_id=' . self::BLOG_ID . '&#038;order_id=' . 0 . '">Manage the order</a> with the app.',
+			$mobile_message
+		);
```
Instead of asserting the exact string value, could this just test whether the return is a _string_ or _null_? That way the wording could change in the future without having to also update the unit test, while still being a useful test.

coreymckrill commented on plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php:
```diff
@@ -0,0 +1,151 @@
+		$this->assertEquals(
+			'Process your orders on the go. <a href="https://woocommerce.com/mobile/">Get the app</a>. Powered by Jetpack.',
+			$mobile_message
+		);
```
This one would be harder to test without specifying the exact wording. One idea might be to look for the substring `href="https://woocommerce.com/mobile/"` since, if I'm not mistaken, that particular URL is only used when no mobile usage is detected.

wzieba committed:
docs: update return types comments

wzieba committed:
test: use wp-specific setup and teardown methods

wzieba committed:
feat: update deep links format

wzieba committed:
test: verify if message contains correct deep link

wzieba committed:
Merge branch 'trunk' into feature/footer_with_mobile_deep_link

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+/**
+ * WC_Mobile_Messaging_Handler
+ */
+class WC_Mobile_Messaging_Handler {
```
Thinking of [this guidance](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/README.md):

> Ideally, all the new code for WooCommerce should consist of classes following the PSR-4 naming and living in this directory, and the code in [the `includes` directory](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/includes/README.md) should receive the minimum amount of changes required for bug fixing.

What do you think of moving this to the `src/` directory and to the `Automattic\WooCommerce` namespace? Additionally, if we do not intend to expose this as part of our public API, we could place it within the `Internal` sub-namespace.

coreymckrill replied:
Ah, I wasn't aware, or at least had forgotten about that guidance. I guess it makes sense for these functions since they're encapsulated as a class. 👍

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+ */
+class WC_Mobile_Messaging_Handler {
+
+	const OPEN_ORDER_INTERVAL_DAYS = 30;
```
So far, we only use this from within this class. We could consider marking it as private (unless you anticipate referencing it from elsewhere in the near future, of course):

```suggestion
	private const OPEN_ORDER_INTERVAL_DAYS = 30;
```

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+	 * @param ?int     $blog_id  of blog to make a deep link for.
+	 * @param DateTime $now      current DateTime.
+	 *
+	 * @return string|null
```
For consistency, what do you think of either following this convention (to match what we use a few lines above, in the case of `@param ?int $blog_id` and in the method signature itself)?

```suggestion
	 * @return ?string
```

Or indeed the reverse: use the `string|null` and `int|null` "union type" style within docblocks? ... Not vital we change this but it strikes me as being more readable when we are consistent.

wzieba replied:
Yes, I completely agree. I've decided to stick with `?string` as it's more compact.

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+	const OPEN_ORDER_INTERVAL_DAYS = 30;
+
+	/**
+	 * Prepares mobile messaging with a deep link
```
```suggestion
	 * Prepares mobile messaging with a deep link.
```

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+	 * Prepares mobile messaging with a deep link
+	 *
+	 * @param WC_Order $order order that mobile message is created for.
+	 * @param ?int     $blog_id  of blog to make a deep link for.
```
I was initially just a little confused by [line 35](https://github.com/woocommerce/woocommerce/pull/34467/files#diff-198890d4431c1f21d2d0a9d3ee3b9d8846cd83ec80bf2b028032459158ff7b1aR35):

```
$has_jetpack            = null !== $blog_id;
```

As it made me wonder what the blog ID has to do with whether Jetpack is present. Now I've read `WC_Email::mobile_messaging()` it's a little clearer, but maybe we could also make a small amendment here:

```suggestion
	 * @param ?int     $blog_id  of blog to make a deep link for (will be null if Jetpack is not enabled).
```

wzieba replied:
Yes, that was indeed confusing. Thanks, I've added the suggested clarification.

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+	 */
+	public static function accept_payment_message(): string {
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%s">Collect payments easily</a> from your customers anywhere with our mobile app.',
+					'woocommerce'
+				)
+			),
+			'https://woocommerce.com/mobile/payments/'
```
Something we could consider:

```suggestion
			esc_html(
				/* translators: 1: opening link tag 2: closing link tag. */
				__(
					'%1$sCollect payments easily%2$s from your customers anywhere with our mobile app.',
					'woocommerce'
				)
			),
			'<a href="https://woocommerce.com/mobile/payments/">',
			'</a>'
```

- `esc_html()` incurs less overhead, and is also stricter
- Moves the HTML out of the translatable string (better for the translator, safer for us)
- Fixes the indentation of the translator note

coreymckrill replied:
I was under the impression that this was harder for translators to deal with, but it there's already precedent elsewhere in the codebase, sounds good to me. If we go this route, it could be simplified further to:


```suggestion
			/* translators: 1: opening link tag 2: closing link tag. */
			esc_html__(
				'%1$sCollect payments easily%2$s from your customers anywhere with our mobile app.',
				'woocommerce'
			),
			'<a href="https://woocommerce.com/mobile/payments/">',
			'</a>'
```

barryhughes replied:
There could well be differing opinions on the matter, though it's definitely not without precedent (one of many cases that can be found in GlotPress):

<img width="551" alt="translating-links" src="https://user-images.githubusercontent.com/3594411/190009048-8115a0c9-94a8-4f09-96f1-8e287e2a493f.png">

I'd push for this strategy because:

- We can add classes, change the target attribute, etc, without altering the translatable string (and without having to wait on translators submitting updates).
- It lets us use stricter escaping, and so can help prevent malicious translation attacks.

> If we go this route, it could be simplified further to...

:100:

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+		);
+
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%s">Manage the order</a> with the app.',
+					'woocommerce'
+				)
+			),
+			esc_url( $deep_link_url )
```
Similar to an earlier note, we could consider:

```suggestion
			esc_html(
				/* translators: 1: opening link tag 2: closing link tag. */
				__(
					'%1$sManage the order%2$s with the app.',
					'woocommerce'
				)
			),
			'<a href="' . esc_url( $deep_link_url ) . '">',
			'</a>'
```

barryhughes commented on plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php:
```diff
@@ -0,0 +1,160 @@
+	 */
+	public static function no_app_message(): string {
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'Process your orders on the go. <a href="%s">Get the app</a>. Powered by Jetpack.',
+					'woocommerce'
+				)
+			),
+			'https://woocommerce.com/mobile/'
```
Similar to earlier notes, how about:

```suggestion
			esc_html(
				/* translators: 1: opening link tag 2: closing link tag. */
				__(
					'Process your orders on the go. %1$sGet the app%2$s. Powered by Jetpack.',
					'woocommerce'
				)
			),
			'<a href="https://woocommerce.com/mobile/">',
			'</a>'
```

barryhughes commented on plugins/woocommerce/includes/emails/class-wc-email.php:
```diff
@@ -256,6 +256,27 @@ public function __construct() {
+	 * Add mobile messaging.
+	 */
+	public function mobile_messaging() {
+		if ( ! $this instanceof WC_Email_New_Order ) {
```
Curious, why not move this method (and the matching `add_action()` call) to `WC_Email_New_Order`? 

Just thinking that this code—which is specific to that other class and is not applicable to all emails—would then be grouped with other code also specific to new order emails.

wzieba replied:
Initially, I wasn't sure if "New Order" was the only type of mail to which we wanted to add messages to. But as it's clear now, I'll move this method and `add_action` directly to `WC_Email_New_Order`, thanks!

barryhughes commented on plugins/woocommerce/includes/wc-ipp-functions.php:
```diff
@@ -0,0 +1,57 @@
+<?php
```
As earlier, we generally don't want to add new code to `includes/`. We can make exceptions of course: but do you think there might be a case here for setting this up as a set of static methods belonging to a new class in our `Automattic\WooCommerce` namespace/the `Internal` sub-namespace)?

Since that would also give us the advantage of autoloading, we could ditch [this line](https://github.com/woocommerce/woocommerce/pull/34467/files#diff-0095ea46d160c94a1276a0eed1d86a5dbc91387c39ac46aae6abd0a1d1d314b9R35).

barryhughes commented on plugins/woocommerce/includes/wc-ipp-functions.php:
```diff
@@ -0,0 +1,57 @@
+
+	$order_has_no_subscription_products = true;
+	foreach ( $order->get_items() as $item ) {
+		$product = $item->get_product();
+
+		if ( $product->is_type( 'subscription' ) ) {
```
Something to consider:

```suggestion
		$product = $item->get_product();

		if ( is_object( $product ) && $product->is_type( 'subscription' ) ) {
```

Though it is presumably unusual, `get_items()` is documented as possibly returning `bool`.

wzieba replied:
Just to better understand: if `get_product()` would return `bool`, the `is_type` executed on it would cause crash, right?

barryhughes replied:
Exactly ... we'd get a fatal error. 

I don't know for sure that there is a chain of events that would lead to us actually receiving `(bool) false` at this point—but the *potential* is there, so being defensive is something to consider.

wzieba committed:
fix: move classes to src/Internal

wzieba committed:
docs: improve PHPDoc comments

wzieba committed:
fix: add visibility modfirier for private-only const

wzieba committed:
fix: improve preparing formatted messages

wzieba committed:
fix: check if order's product is an object

wzieba committed:
fix: move mobile messaging directly to new order email

wzieba committed:
feat: add blog id to mobile payments deep link

wzieba commented:
In 5029618 I've added `blog_id` parameter to the deep link for tracking reasons (internal communication: p6JqRr-7ou-p2). cc: @toupper

toupper commented:
Thanks for addressing the tracking as well @wzieba :) On p6JqRr-7ou-p2 we discuss about the best way to pass the values we want to track through the URL. Since we already track `utm_source` and `utm_content` (please note that they don't track blog_id but blog_url which is not related to our case here), I thought that we can use those so the alpha team doesn't have to change anything. In that case, the parameter name for the `blog_id` would be `utm_source`. e.g `woocommerce.com/mobile/orders/details?blog_id=123&order_id=345&utm_source=mobile_deeplink_orders&utm_content=123456 `being 123456 the blog_id. Please consider that when implementing the tracking parameters :)

wzieba commented:
Thanks @toupper ! Sure, good idea - I'll update the code to meet those requirements.

So heads up @barryhughes and @coreymckrill - I'm now working on a few improvements. I'll re-request your review when it'll be ready. Thank you!

barryhughes commented:
Sounds good, thanks for the work so far!

> After moving files, I broke the feature. When creating a new Order, I have this: [error]

Following up on this one, in the first instance it is probably worth running `composer dump` (if you haven't already tried doing so), to update the classmaps.

wzieba committed:
feat: add tracking query arguments

wzieba committed:
feat: update copy selection logic

wzieba committed:
Merge branch 'trunk' into feature/footer_with_mobile_deep_link

wzieba committed:
fix: do not crash when Jetpack's blog id is null

wzieba commented:
@toupper 👋
In d8cf172 I've added `utm_source` and `utm_content` query parameters. `utm_source` can be:
- `mobile_deeplink_payments`
- `mobile_deeplink_orders_details`
- `mobile_deeplink_no_app`

and `utm_content` will be `blog_id` (or `0` if site doesn't have Jetpack). I think it meets the requirements discussed in p6JqRr-7ou-p2, right?

----

@barryhughes @coreymckrill 👋

In addition to the above-described change, in 6ee6341 I've updated the logic for selecting the correct copy. I've also updated the diagram in the PR description to spot the difference quickly. Ready for another round of review!

> (...) in the first instance it is probably worth running composer dump

It worked like a charm, thanks!

coreymckrill reviewed (approved):
👍 Nice work getting this put together and including unit tests!
